### ISSUE TITLE
Option to filter diagnostics based on version control status

### ISSUE THREAD

**Issue by jroelofs** [clangd/clangd#822]
Clang-tidy diagnostics are pretty noisy when working on llvm itself, so it would be helpful to have an option in the extension to filter them by whether they're near changed lines in the source. This would make the remaining diagnostics much more actionable and likely to be fixed, IMO.

**Comment by jroelofs** [comment]
(vscode extension, that is)

**Comment by sam-mccall** [comment]
The idea of only showing warnings on changed lines per VCS is definitely interesting - roughly, in a warning-infested codebase you may choose simply to not introduce new ones. A couple of hesitations:

 - the the warning location isn't always the same as the modification that triggers it. Simple example is an "unused variable" warning appears on the (unchanged) declaration line after you remove the usage. Defining "near" the changed lines gets messy quickly.
 - Clangd doesn't know anything about the baseline, as this isn't part of LSP. While it may be possible to build it from scratch, there's a lot of complexity here: different VCSes, the "correct" baseline for different workflows etc. Editors (or at least VSCode) have a model of this already. It's tempting to say this should be a feature/extension at the editor level instead (hiding warnings on unchanged lines).

**Comment by jroelofs** [comment]
Good point re: unused variable warnings. Maybe the solution to that is to build with/without the local changes, and remove diagnostics that are in the intersection. Source locations can't be compared directly, of course, but a lookup table could probably be built from the diff itself.

Does the VSCode extension have its own bug tracker?

**Comment by HighCommander4** [comment]
> Does the VSCode extension have its own bug tracker?

Yup, https://github.com/clangd/vscode-clangd

**Comment by sam-mccall** [comment]
Though I did mean a separate feature/extension, in the end the feature doesn't seem specific to clangd.

Building the baseline revision to compare diagnostics is tempting, but I'm not sure how feasible it is, we'd need all the baseline sources (headers) and in principle the compile commands too. Not impossible but certainly complicated!
